language-plutus-core/src/Language/PlutusCore/Core/Instance/Eq.hs

```diff
@@ -52,17 +52,148 @@ instance (HasUniques (Program tyname name uni ann), GEq uni, Closed uni, uni `Ev
+
+    f x == (\x -> f x x)
+
+we can't simply cut @(\x -> f x x)@ into @f x@, because on the LHS we have a type that references
+@x@ (a completely different free variable with the same name that we have bound on the LHS). This
+is similar to the aforementioned problem where we could not eta-expand @f x@ into @\x -> f x x@.
+
+Hence we need to explicitly record that @x@ is not in scope in @f x@ on the RHS and fail on
+encountering it while checking
+
+    f x == f x
+
+after "eta-contracting" @\x -> f x x@ into the wrong @f x@.
+
+But our problems don't end here. Consider the following equation:
+
+    f == (\x x -> f x x)
+
+If we simply cut @\x x -> f x x@ into @\x -> f x@ and record that @x@ must be out of scope in that
+type and proceed, then we'll find out that @x@ is indeed out of scope there, because there's a
+lambda binding the same variable and thus shadowing the original one. And then we proceed and
+eta-contract @\x -> f x@ into @f@. Clearly those two types are very different and the algorithm
+does not handle this case correctly.
+
+Note however that we do want to handle shadowing like that when a lambda that binds a variable
+shadowing an existing one is not amongst the leading lambdas, i.e. it's either applied in the body
+of the n-ary lambda term or an argument there. For example, when performing this equality
+check:
+
+    f (\x -> x) == (\x -> f (\x -> x) x)
+
+we do want the first @x@ bound on the RHS to be shadowed by the second one.
+
+What we can do here is remove all the leading lambdas in a single pass and ensure that the
+variables they bind are all distinct (as @\x x -> f@ can never be eta-contracted into @f@
+regardless of what @f@ is).
+
+This wouldn't be a correct thing to do, if we were to eta-contract both the sides of an equation
+first and only then check their equality, because then we wouldn't be able to eta-contract the RHS
+of this equality check that is meant to succeed:
+
+    (\x -> f) == (\x x -> f x)
+
+However the case where both the sides start with a lambda is already correctly handled by the main
+machinery, so when we do eta-contraction, we can safely assume that the other side of the equation
+does not have any leading lambdas.
+-}
+
+-- | Eta-contract a type and feed the result to the provided continuation that gets executed in a
+-- context extended with all removed lambda-bound variables mapped to 'Nothing' (meaning they are
+-- not in scope in the eta-contracted function).
+withFullyEtaContracted
+    :: ( HasUnique tyname TypeUnique
+       , HasMapping (Bilateral ren) (dir TypeUnique) (Maybe TypeUnique)
+       )
+    => (TypeUnique -> dir TypeUnique)
+    -> tyname
+    -> Type tyname uni ann
+    -> (Type tyname uni ann -> EqRename ren)
+    -> EqRename ren
+withFullyEtaContracted dir name0 = stripTyLams [name0 ^. unique] where
+    stripTyLams uniqs (TyLam _ name _ body') k
+        | uniq `elem` uniqs = empty
+        | otherwise         = stripTyLams (uniq : uniqs) body' k
```

Comment:
Do we need to accumulate all the uniques like this? You should be able to do one-level only eta contraction (i.e. a lambda immediately around an application) provided you start from the inside, i.e. `\x -> \y -> f x y ==> \x -> f x ==> f`. Maybe this doesn't work with the continuation-based approach, I haven't tried it, but it seems more "obvious" to me, since you only have to think about two adjacent levels of syntax.

Reply:
So my only objection to that is "that doesn't sound efficient at all".

Reply:
I don't see why not? You call eta-contract on all the subterms, then you check whether the top-level is the pattern you want, and then either contract or not. Laziness should still stop it from traversing more of the tree than it needs to.

Reply:
> I don't see why not? You call eta-contract on all the subterms, then you check whether the top-level is the pattern you want, and then either contract or not.

Best case implementation would be O(n) where `n` is the size of the whole type (because you need to know all the free variable for deciding if eta-contraction is possible). The current implementation is O(n^2) where `n` is the number of leading lambdas.